### ISSUE TITLE
Typescript definition file for the signalr-client.js libary [User Experience]

### ISSUE THREAD

**Issue by dotnetnoobie** [aspnet/SignalR#239]
Is is possible to get a signalr-client.d.ts file

I can get the .d.ts files for all the components/parts of the signalr client

but I would like to be able to write the typescript file that produces the js for my page where I use signalr

when I try to reference the signalr-client.js into the .ts file I have mixed results and have only kind of made it work, but not really only partly.  It looks like the signar-client-js is setup to produce module/exports which I have read about on the typescript site

I am a typescript novice, so I might be just doing it wrong?
Or is it possible to get a signalr-client.d.ts produced like the rest of the components of the client can be?

**Comment by muratg** [comment]
@moozzyk I think this may be done already...

**Comment by moozzyk** [comment]
Partially. I need to play a bit to see if it fits the bill

**Comment by muratg** [comment]
.d.ts files are created now. In the next milestone, we plan to push the typings to npm `typings` scope.

**Comment by valeriob** [comment]
hi @muratg 
i've compiled the project and i see that many .d.ts files are created. Do you plan to create a single signalr-client.d.ts to use with the corrisponding dist\browser\signalr-client.js ?

**Comment by muratg** [comment]
@valeriob yes, we plan to have a single .d.ts file for the client library.

**Comment by malichishti** [comment]
@muratg Thank you for the update. May i know when can one expect the typings to be available via npm?

Regards,
Ali

**Comment by moozzyk** [comment]
At the moment they are in the module. Just `import * as SignalR from "@aspnet/signalr-client"` and you should be good to go.

**Comment by malichishti** [comment]
@moozzyk I cant find that module in npm?

**Comment by malichishti** [comment]
If i use [signalr-client](https://www.npmjs.com/package/signalr-client) package from npm and import as below:
`import * as SignalR from "signalr-client"`
I get:
![image](https://user-images.githubusercontent.com/17155996/30414478-2bb92ad4-9967-11e7-818d-6ecc7c6fc96c.png)
when called as below:
 
![image](https://user-images.githubusercontent.com/17155996/30414501-43bc8360-9967-11e7-8746-bc56953d2ff0.png)

I think it is coming from 
![image](https://user-images.githubusercontent.com/17155996/30414524-61e0e318-9967-11e7-8a67-c7cf6d5ad8e1.png)

**Comment by moozzyk** [comment]
XmlHttpRequest and WebSockets need to be polyfilled. I used xmlhttprequest and websocket packages and stuck them on the `global` object.

**Comment by BrennanConroy** [comment]
@malichishti That isn't our package. Look at the [README](https://github.com/aspnet/SignalR/blob/dev/README.md) for instructions on how to get our npm module.

**Comment by davidfowl** [comment]
@moozzyk is this pretty much done now?

**Comment by moozzyk** [comment]
I think we can close it for now. We might still consider putting type definitions to definitely typed if people asked for it.

**Comment by davidfowl** [comment]
`node_modules\@aspnet\signalr-client\dist\src\index.d.ts`

**Comment by moozzyk** [comment]
@valeriob - I am not sure why you need a single file? I wrote a [blog pos](https://blog.3d-logic.com/2017/09/18/the-signalr-for-asp-net-core-javascript-client-part-1-web-applications/)t about using the client from web applications which included a TypeScript client. I did not have any issues with using typing from the npm even though they are not in just one file.

**Comment by davidfowl** [comment]
Oh, you mean a single file with all of the definitions inlined?

**Comment by valeriob** [comment]
1) because it's fine for simple apps, but you do not want your typescript compilation to slow down due to scanning the whole node folder (that is huge) just to find a simple type definition file.
I'll try to configure tsconfig.json to pick those up with some files or include.
2) because you force ppl to use a typescript module system just to load 10k of javascript client. If you want to use namespaces or nothing at all you can't.

**Comment by davidfowl** [comment]
@valeriob can you point us to other examples on npm that do this?

**Comment by valeriob** [comment]
Sure : 
https://github.com/BabylonJS/Babylon.js/tree/master/dist or everything in DefinitelyTyped.
I mean not everyone builds spa or want to mess with unstable/ugly/volatile client build/module systems 😄

**Comment by moozzyk** [comment]
If we create a single file typings it would not be part of the signalr-client module but would live in the DefinitelyTyped

**Comment by davidfowl** [comment]
> or everything in DefinitelyTyped.

DefinitelyTyped is dead though. 

We're not pushing back, we're just looking for guidance. Does angular do the same?

**Comment by valeriob** [comment]
What about everyone who does not use angular or build spa ?

**Comment by moozzyk** [comment]
I meant whatever the @types scope uses.

**Comment by davidfowl** [comment]
@valeriob what kind of application are you building?

**Comment by valeriob** [comment]
AspNetCore server rendered web applications.
Where just same pages use advanced binding frameworks like knockout or vuejs written in typescript, but without any module system, because it's not needed.

**Comment by davidfowl** [comment]
@valeriob any chance to have an example of what that looks like on github? I'm curious how you're consuming other libraries.

PS: This is purely educational for me

**Comment by davidfowl** [comment]
@moozzyk yes hat makes sense. It seems like its just another package. Let's just open a new bug specifically about the types scope

**Comment by moozzyk** [comment]
https://github.com/aspnet/SignalR/issues/951

**Comment by muratg** [comment]
@davidfowl the way you consume?

package.json:
```json
{
  "name": "WebApp",
  "private": true,
  "version": "0.1.0",
  "devDependencies": {
    "isomorphic-fetch": "^2.2.1",
    "@types/isomorphic-fetch": "0.0.34"
  }
}

```

Some libraries bring in their typings within the package and have a `@types` scoped package which just says "you don't need this" on `npm install`.

**Comment by valeriob** [comment]
thanks @davidfowl this is educational for me too.
We just develop aspnet(core) applications (well organized in areas and feature folders wich contains anything from controller, correlated views and typescript), and just a few important pages are "mini-spa" where the use case requires interaction, reactivity etc...
So we do not buy in the whole spa dependency madness that is going on, just build a view model with some framework (knockout/vue/newFancy) and be done with it.
If we want to change a framework we just migrate one page at time (already done two times in few years, jquery, knockout, vue1,vue2 😄 ), there are no dependencies between each other (maybe some typescript).

I tried to build each mini spa with web-pack (and typescript plugin with modules etc), but it was slow as hell  (because each run is a typescript compilation)! So the simple thing works and it is lightning fast because typescript compilation happens once, and you just need a dumb blundler to chain files together for the single mini spa.
Another benefit is that you have to explicitly declare dependency in any use case so your bundle config file becomes a manifest of your dependencies and you do not have to find and grep in the code base to look for places where you used a framework.
I do not know if it's the right way, but this model has work very well for us for few years.

**Comment by mnmr** [comment]
If I may chime in on this: we've built our entire application using a single namespace, specifically to avoid dealing with modules and imports (this was before rollup/browserify came along to make it feasible to use modules for client-side logic without killing your page load times).

Most (if not all) of the typings we used came from DefinitelyTyped and allowed direct use through either a global variable or by extending e.g. jQuery. It should be possible to support both module-less and module-based code with a single definition file. However, the current index.d.ts shipping with the SignalR client requires you to use the import statement, which the ts compiler denies when you've set the "module": "none" option in tsconfig.json.

I've extracted a few bits from the @types for fastclick to illustrate how they solve it:

`// Type definitions for FastClick v1.0.3
// Project: https://github.com/ftlabs/fastclick

interface FastClickObject {
}
interface FastClickOptions {
}

interface FastClickStatic {
	new(layer: any, options?: FastClickOptions): FastClickObject;
}

declare module "fastclick" {
	function fastclick(layer: any, options?: FastClickOptions): FastClickObject;
	namespace fastclick {
		var FastClick: FastClickStatic;
	}
	export = fastclick;
}

declare var FastClick: FastClickStatic;
`

The trick here seems to be to declare a FastClickStatic variable. This allows module-less people to access the exports of the module, but without using modules per se.
